### DISCUSSION TITLE
Building lapack fails in Windows11

### DISCUSSION THREAD

**Discussion by wknm1997** [scivision/mumps-superbuild#53] [Q&A] [answered]
Dear all,
Thanks for providing the useful cmake files to help build mumps in Windows. I want to build lapack via cmake according to the description. But I found out, should the `cmake --build build -t scalapack` here be `cmake --build scripts/build -t scalapack` ?
After I ran this command, unfortunately, it gave the following error :
`-- Could NOT find MPI_C (missing: MPI_C_LIB_NAMES MPI_C_WORKS)
-- Could NOT find MPI_Fortran (missing: MPI_Fortran_LIB_NAMES MPI_Fortran_WORKS)
CMake Error at D:/Program Files/CMake/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find MPI (missing: MPI_C_FOUND MPI_Fortran_FOUND C Fortran)
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  D:/Program Files/CMake/share/cmake-3.25/Modules/FindMPI.cmake:1837 (find_package_handle_standard_args)
  CMakeLists.txt:25 (find_package)

-- Configuring incomplete, errors occurred!
See also "D:/Program Files/MATLAB/R2022b/bin/MUMPS_5.5.1/scripts/build/scalapack-prefix/src/scalapack-build/CMakeFiles/CMakeOutput.log".
See also "D:/Program Files/MATLAB/R2022b/bin/MUMPS_5.5.1/scripts/build/scalapack-prefix/src/scalapack-build/CMakeFiles/CMakeError.log".
CMakeFiles\scalapack.dir\build.make:90: recipe for target 'scalapack-prefix/src/scalapack-stamp/scalapack-configure' failed
mingw32-make.exe[3]: *** [scalapack-prefix/src/scalapack-stamp/scalapack-configure] Error 1
CMakeFiles\Makefile2:113: recipe for target 'CMakeFiles/scalapack.dir/all' failed
mingw32-make.exe[2]: *** [CMakeFiles/scalapack.dir/all] Error 2
CMakeFiles\Makefile2:120: recipe for target 'CMakeFiles/scalapack.dir/rule' failed
mingw32-make.exe[1]: *** [CMakeFiles/scalapack.dir/rule] Error 2
Makefile:136: recipe for target 'scalapack' failed
mingw32-make.exe: *** [scalapack] Error 2`

I would like to say, I have Intel Parallel Studio2020 MPI and Microsoft MPI. Also, I added their folders to the system environment variables. But still this error has not changed. So I sincerely ask you, is there something wrong with my operation?

thanks

**Comment by wknm1997** [comment]
Yesterday, I was using -G "MinGW Makefiles". Today I realized to build Lapack with VS2019+oneAPI instead. But it throws a new error.
`  Microsoft (R) Windows (R) Resource Compiler Version 10.0.10011.16384`
`  Copyright (C) Microsoft Corporation.  All rights reserved.`
`  Linking...`
`LINK : fatal error LNK1561: 必须定义入口点 [D:\Program Files\MATLAB\lib\MUMPS_5.5.1\scripts\build\lapack.vcxproj]`
` `
`  Build log written to  "file://D:/Program%20Files/MATLAB/lib/MUMPS_5.5.1/scripts/build/lapack-prefix/src/lapack-build/`
`  run_test_zcomplexabs.dir/Debug/BuildLog.htm"`
`  run_test_zcomplexabs - 1 error(s), 0 warning(s)`
`LINK : fatal error LNK1561: 必须定义入口点 [D:\Program Files\MATLAB\lib\MUMPS_5.5.1\scripts\build\lapack.vcxproj]`
` `
 ` Build log written to  "file://D:/Program%20Files/MATLAB/lib/MUMPS_5.5.1/scripts/build/lapack-prefix/src/lapack-build/`
`  run_test_zminMax.dir/Debug/BuildLog.htm"`
`  run_test_zminMax - 1 error(s), 0 warning(s)`
`LINK : fatal error LNK1561: 必须定义入口点 [D:\Program Files\MATLAB\lib\MUMPS_5.5.1\scripts\build\lapack.vcxproj]`
` `
 ` Build log written to  "file://D:/Program%20Files/MATLAB/lib/MUMPS_5.5.1/scripts/build/lapack-prefix/src/lapack-build/`
`  run_test_zcomplexmult.dir/Debug/BuildLog.htm"`
`  run_test_zcomplexmult - 1 error(s), 0 warning(s)`
`LINK : fatal error LNK1561: 必须定义入口点 [D:\Program Files\MATLAB\lib\MUMPS_5.5.1\scripts\build\lapack.vcxproj]`
` `
 ` Build log written to  "file://D:/Program%20Files/MATLAB/lib/MUMPS_5.5.1/scripts/build/lapack-prefix/src/lapack-build/`
`  run_test_zcomplexdiv.dir/Debug/BuildLog.htm"`
`  run_test_zcomplexdiv - 1 error(s), 0 warning(s)`

Part of the error report is shown here, and I've found that this may have something to do with not specifying "main", but this is a library file. So maybe it has something to do with some bug in the cmake file?

**Reply by scivision** [reply]
With Intel oneAPI or Parallel Studio, you should not build Lapack and Scalapack. They are included with Intel HPC toolkit. The build of Lapack and Scalapack will not be successful.

**Comment by wknm1997** [comment]
Hello, after many days of hard work, I found that although your project can be successfully built for the linux system. But there are still many bugs for Windows system. One of them is that when compiling (cdsz)mumps, the following error will be reported:
```
dmumps.lib(mumps_c.obj) : error LNK2019: unresolved external symbol DMUMPS_F77, referenced in function dmumps_c
dmumps.lib(mumps_c.obj) : error LNK2019: unresolved external symbol DMUMPS_SET_TMP_PTR, referenced in function DMUMPS_SET_TMP_PTR_C
```
I found that this is due to the 190 and 245 lines of the mumps_c.c file, and there is an error in the macro definition of the MUMPS_F77 and MUMPS_SET_TMP_PTR functions. Maybe there should be {} instead of ;.
I have to say that after I changed this symbol, the above error did not appear again when compiling. Of course, in this process, there are more than one mistakes in the project. If you are interested, we can communicate.

**Reply by scivision** [reply]
On Windows, you must use -G Ninja or -G "MinGW Makefiles". The Visual Studio generator is known not to work.

**Comment by scivision** [comment, marked as answer]
I made a separate [Windows readme](https://github.com/scivision/mumps/blob/main/Readme_Windows.md)

**Reply by wknm1997** [reply]
Thank you for your effort on my mistake. But when I follow your new method, I still get the error that lapack cannot be found. When I used cmakegui to view this error, I found that the mkl library was recognized. The error report is as follows:
```sh
The C compiler identification is GNU 6.3.0
The Fortran compiler identification is GNU 6.3.0
Detecting C compiler ABI info
Detecting C compiler ABI info - done
Check for working C compiler: D:/mingw64/mingw64/bin/gcc.exe - skipped
Detecting C compile features
Detecting C compile features - done
Detecting Fortran compiler ABI info
Detecting Fortran compiler ABI info - done
Check for working Fortran compiler: D:/mingw64/mingw64/bin/gfortran.exe - skipped
MUMPS 5.5.1.9  CMake 3.25.1  Toolchain 
checking that C and Fortran compilers can link
checking that C and Fortran compilers can link - OK
Performing Test CMAKE_HAVE_LIBC_PTHREAD
Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
Looking for pthread_create in pthreads
Looking for pthread_create in pthreads - not found
Looking for pthread_create in pthread
Looking for pthread_create in pthread - found
Found Threads: TRUE  
Performing Test LAPACK_s_FOUND
Performing Test LAPACK_s_FOUND - Failed
Performing Test LAPACK_d_FOUND
Performing Test LAPACK_d_FOUND - Failed
CMake Error at D:/Program Files/CMake/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find LAPACK (missing: LAPACK_links)
Call Stack (most recent call first):
  D:/Program Files/CMake/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindLAPACK.cmake:483 (find_package_handle_standard_args)
  cmake/lapack.cmake:22 (find_package)
  CMakeLists.txt:39 (include)


Configuring incomplete, errors occurred!
See also "C:/Users/21478/Desktop/mumps/build/CMakeFiles/CMakeOutput.log".
See also "C:/Users/21478/Desktop/mumps/build/CMakeFiles/CMakeError.log".
```
![JGN}AI9TIM0$8(@BP6P1KNE](https://user-images.githubusercontent.com/85986757/212671042-cfb08026-3c2d-41fc-8924-25acd05fe4ab.png)

**Reply by scivision** [reply]
MKL supports GCC only on Linux. On Windows to use MKL you need to use the oneAPI compiler too.

**Reply by wknm1997** [reply]
thank you for your patience. But I still didn't succeed, I used mingw and ninja, but I got .a instead of .lib. During this process, there was still not found lapack, so I manually modified the cache file and let it lock ~/mylibs/libs/lapack.lib.
So I wonder if I should use oneAPI instead of GNU as my compiler. But my cmake defaults to GNU. So I used the following command.
`cmake -B build -G "MinGW Makefiles" -DCMAKE_C_COMPILER=D:/Program/Intel/oneAPI/compiler/latest/windows/bin/icx.exe -DCMAKE_CXX_COMPILER=D:/Program/Intel/oneAPI/compiler/latest/windows/bin/icx.exe -DCMAKE_FORTRAN_COMPILER=D:/Program/Intel/oneAPI/compiler/latest/windows/bin/ifx.exe`
The obtained cmakeError.log is as follows
```sh
Checking whether the Fortran compiler is Compaq using "-what" did not match "Compaq Visual Fortran":
Intel(R) Fortran Compiler for applications running on Intel(R) 64, Version 2023.0.0 Build 20221201
Copyright (C) 1985-2022 Intel Corporation. All rights reserved.

ifx: command line error: no files specified; for help type "ifx /help"
Checking whether the Fortran compiler is NAG using "-V" did not match "NAG Fortran Compiler":
Intel(R) Fortran Compiler for applications running on Intel(R) 64, Version 2023.0.0 Build 20221201
Copyright (C) 1985-2022 Intel Corporation. All rights reserved.

ifx: command line warning #10155: ignoring option '/V'; argument required
ifx: command line error: no files specified; for help type "ifx /help"
Checking whether the Fortran compiler is Compaq using "-what" did not match "Compaq Visual Fortran":
Intel(R) Fortran Compiler for applications running on Intel(R) 64, Version 2023.0.0 Build 20221201
Copyright (C) 1985-2022 Intel Corporation. All rights reserved.

ifx: command line error: no files specified; for help type "ifx /help"
Checking whether the Fortran compiler is NAG using "-V" did not match "NAG Fortran Compiler":
Intel(R) Fortran Compiler for applications running on Intel(R) 64, Version 2023.0.0 Build 20221201
Copyright (C) 1985-2022 Intel Corporation. All rights reserved.

ifx: command line warning #10155: ignoring option '/V'; argument required
ifx: command line error: no files specified; for help type "ifx /help"
Detecting C compiler ABI info failed to compile with the following output:
Change Dir: C:/Users/21478/Desktop/mumps/build/CMakeFiles/CMakeScratch/TryCompile-rovcfv

Run Build Command(s):D:/mingw64/mingw64/bin/mingw32-make.exe -f Makefile cmTC_12145/fast && D:/mingw64/mingw64/bin/mingw32-make.exe  -f CMakeFiles\cmTC_12145.dir\build.make CMakeFiles/cmTC_12145.dir/build
mingw32-make.exe[1]: Entering directory 'C:/Users/21478/Desktop/mumps/build/CMakeFiles/CMakeScratch/TryCompile-rovcfv'
Building C object CMakeFiles/cmTC_12145.dir/CMakeCCompilerABI.c.obj
D:\Program\Intel\oneAPI\compiler\latest\windows\bin\icx.exe  /nologo   /DWIN32 /D_WINDOWS  /Ob0 /Od /RTC1 -MDd -Zi /FoCMakeFiles\cmTC_12145.dir\CMakeCCompilerABI.c.obj /FdCMakeFiles\cmTC_12145.dir/ -c "D:\Program Files\CMake\share\cmake-3.25\Modules\CMakeCCompilerABI.c"
Linking C executable cmTC_12145.exe
"D:\Program Files\CMake\bin\cmake.exe" -E cmake_link_script CMakeFiles\cmTC_12145.dir\link.txt --verbose=1
"D:\Program Files\CMake\bin\cmake.exe" -E vs_link_exe --intdir=CMakeFiles\cmTC_12145.dir --rc=rc --mt=CMAKE_MT-NOTFOUND --manifests -- D:\Program\Intel\oneAPI\compiler\latest\windows\bin\icx.exe /nologo @CMakeFiles\cmTC_12145.dir\objects1  /Qoption,link,/machine:x64  /debug /INCREMENTAL /Qoption,link,/subsystem:console /link /out:cmTC_12145.exe /implib:cmTC_12145.lib /pdb:C:\Users\21478\Desktop\mumps\build\CMakeFiles\CMakeScratch\TryCompile-rovcfv\cmTC_12145.pdb /version:0.0 
RC Pass 1: command "rc /fo CMakeFiles\cmTC_12145.dir/manifest.res CMakeFiles\cmTC_12145.dir/manifest.rc" failed (exit code 0) with the following output:
系统找不到指定的文件。CMakeFiles\cmTC_12145.dir\build.make:98: recipe for target 'cmTC_12145.exe' failed
mingw32-make.exe[1]: *** [cmTC_12145.exe] Error -1
mingw32-make.exe[1]: Leaving directory 'C:/Users/21478/Desktop/mumps/build/CMakeFiles/CMakeScratch/TryCompile-rovcfv'
Makefile:126: recipe for target 'cmTC_12145/fast' failed
mingw32-make.exe: *** [cmTC_12145/fast] Error 2




Determining if the C compiler works failed with the following output:
Change Dir: C:/Users/21478/Desktop/mumps/build/CMakeFiles/CMakeScratch/TryCompile-1o72w6

Run Build Command(s):D:/mingw64/mingw64/bin/mingw32-make.exe -f Makefile cmTC_3079c/fast && D:/mingw64/mingw64/bin/mingw32-make.exe  -f CMakeFiles\cmTC_3079c.dir\build.make CMakeFiles/cmTC_3079c.dir/build
mingw32-make.exe[1]: Entering directory 'C:/Users/21478/Desktop/mumps/build/CMakeFiles/CMakeScratch/TryCompile-1o72w6'
Building C object CMakeFiles/cmTC_3079c.dir/testCCompiler.c.obj
D:\Program\Intel\oneAPI\compiler\latest\windows\bin\icx.exe  /nologo   /DWIN32 /D_WINDOWS  /Ob0 /Od /RTC1 -MDd -Zi /FoCMakeFiles\cmTC_3079c.dir\testCCompiler.c.obj /FdCMakeFiles\cmTC_3079c.dir/ -c C:\Users\21478\Desktop\mumps\build\CMakeFiles\CMakeScratch\TryCompile-1o72w6\testCCompiler.c
Linking C executable cmTC_3079c.exe
"D:\Program Files\CMake\bin\cmake.exe" -E cmake_link_script CMakeFiles\cmTC_3079c.dir\link.txt --verbose=1
"D:\Program Files\CMake\bin\cmake.exe" -E vs_link_exe --intdir=CMakeFiles\cmTC_3079c.dir --rc=rc --mt=CMAKE_MT-NOTFOUND --manifests -- D:\Program\Intel\oneAPI\compiler\latest\windows\bin\icx.exe /nologo @CMakeFiles\cmTC_3079c.dir\objects1  /Qoption,link,/machine:x64  /debug /INCREMENTAL /Qoption,link,/subsystem:console  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib  /link /out:cmTC_3079c.exe /implib:cmTC_3079c.lib /pdb:C:\Users\21478\Desktop\mumps\build\CMakeFiles\CMakeScratch\TryCompile-1o72w6\cmTC_3079c.pdb /version:0.0 
RC Pass 1: command "rc /fo CMakeFiles\cmTC_3079c.dir/manifest.res CMakeFiles\cmTC_3079c.dir/manifest.rc" failed (exit code 0) with the following output:
系统找不到指定的文件。CMakeFiles\cmTC_3079c.dir\build.make:98: recipe for target 'cmTC_3079c.exe' failed
mingw32-make.exe[1]: *** [cmTC_3079c.exe] Error -1
mingw32-make.exe[1]: Leaving directory 'C:/Users/21478/Desktop/mumps/build/CMakeFiles/CMakeScratch/TryCompile-1o72w6'
Makefile:126: recipe for target 'cmTC_3079c/fast' failed
mingw32-make.exe: *** [cmTC_3079c/fast] Error 2
```

**Reply by scivision** [reply]
To use oneAPI on Windows, it's easiest to use the special command prompt for oneAPI. Under Windows Start, Programs, Intel "oneAPI command prompt for Intel 64 for Visual Studio 2022" or just Windows Search "Intel oneAPI command prompt for Intel 64 for Visual Studio 2022" 

If you can't find that special command prompt, you can set the necessary variables to use oneAPI at https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html

This can also be done from PowerShell:

```pwsh
$Env:CC="icx"
$Env:FC="ifx"
$Env:CXX="icx"

cmd.exe /c '"%PROGRAMFILES(X86)%\Intel\oneAPI\setvars.bat" && pwsh'
```

Whenever compiling programs on Windows, the proper environment variables and/or command prompt must be used. This is for any code project on Windows, not just MUMPS.

Use a fresh build directory so that the Intel compilers are properly recognized and not mixed up with the failed GCC build.

**Reply by scivision** [reply]
Then the command line from oneAPI command prompt would be like:

```sh
cmake -G Ninja -B build -DCMAKE_C_COMPILER=icx -DCMAKE_Fortran_COMPILER=ifx -DBUILD_SINGLE=yes -DBUILD_DOUBLE=yes -DBUILD_COMPLEX=yes -DBUILD_COMPLEX16=yes
````

**Reply by wknm1997** [reply]
> Then the command line from oneAPI command prompt would be like:
> 
> ```shell
> cmake -G Ninja -B build -DCMAKE_C_COMPILER=icx -DCMAKE_Fortran_COMPILER=ifx -DBUILD_SINGLE=yes -DBUILD_DOUBLE=yes -DBUILD_COMPLEX=yes -DBUILD_COMPLEX16=yes
> ```

Thank you, with your patient help for so many days, I finally successfully compiled mumps with windows. Thanks for your contribution to the project.I will share my joy with you as soon as possible when my paper is published!

**Comment by wknm1997** [comment]


**Reply by scivision** [reply]
You need to have the correct command prompt and environment variables as above, or even if someone emailed you the binary libraries, the build would still fail. The correct programming environment is essential for building any Windows program.

**Reply by scivision** [reply]
The default build type for MUMPS is Release (optimized)

**Reply by scivision** [reply]
The Matlab interface is from 2006; I'm not sure if it is maintained. The Matlab interface didn't seem to totally work for me. I would work with plain MUMPS.
